### ISSUE TITLE
Webhook signature verification fails

### ISSUE THREAD

**Issue by quintasda** [getsentry/sentry#71025]
### Environment

SaaS (https://sentry.io/)

### Steps to Reproduce

[Verify the Signature](https://docs.sentry.io/organization/integrations/integration-platform/webhooks/#sentry-hook-signature:~:text=comment-,Sentry%2DHook%2DSignature,-This%20header%20represents)

[Customer case](https://www.notion.so/sentry/Webhook-signature-verification-fail-fbf67621bd8c4a61923355698cfa6bfa)

### Expected Result

The signature matches

### Actual Result

The signature doesn't match in ~10% of the messages 

### Product Area

Settings - Integrations

**Comment by Dhrumil-Sentry** [comment]
Thanks for this report, we will investigate this cc @sentaur-athena

**Comment by sentaur-athena** [comment]
This issue is fixed now. We investigated and the reason was a new json serializer we started using since May 9. We reverted the changes to the previous serializer and looking at dashboards the 400 and 401 responses are eliminated.

**Comment by sentaur-athena** [comment]
@quintasda as we're adding back the new serializer wanted to make sure we're not breaking customer's experience. Can you please provide me with code snippet of how you're validating signature?

You mentioned [Verify the Signature](https://docs.sentry.io/organization/integrations/integration-platform/webhooks/#sentry-hook-signature:~:text=comment-,Sentry%2DHook%2DSignature,-This%20header%20represents) from our docs but I was wondering if you're using the python or js example? Also what you use to dump json.

**Comment by jasonyonker** [comment]
@sentaur-athena I'm having issues with this using the exact code from the Sentry docs:

```
    const hmac = crypto.createHmac('sha256', clientSecret);
    hmac.update(JSON.stringify(body), 'utf8');
    const digest = hmac.digest('hex');
    return digest === signature;
```

This is on NodeJS 19.

Frankly, I don't think this solution is technically sound. `JSON.stringify` does not produce stable, deterministic output across different Node versions and even if it did there's no guarantee that serializers for Python, Ruby, Go, etc. will produce the same output. There are dedicated libraries that attempt to do this but I don't know of any standard that is widely supported on different platforms.

See the discussion here: https://github.com/nodejs/node/issues/15628

**Comment by sentaur-athena** [comment]
@jasonyonker I agree with you. We should look into something more stable here. Let me look into it and get back to you.

**Comment by sentaur-athena** [comment]
@jasonyonker I read the discussion you sent and I don't think any string serializer would work here since any customer might use a different language and framework. I'll update the docs and suggest to do the verification on raw bytes of body. The catch with that is that some frameworks don't provide raw body bytes at all. Looks like `fastify` and `express` might be some examples.

Can you try with `hmac.update(raw_request_body)`?

**Comment by Christinarlong** [comment]
Hi, have there been any updates for this issue ?

**Comment by Fwang36** [comment]
hey do we have an ETA on when our docs would be updated?  Support received another request here in this ticket: https://sentry.zendesk.com/agent/tickets/156220

The user in the ticket also copied the node example in our docs but is consistently getting a mismatch.

**Comment by sentaur-athena** [comment]
@Fwang36 It's tricky because based on what package they're using the may or may not have access to raw request body. We will update the docs soon to reflect that.
